### PR TITLE
[MRG] Add vtki 3d backend

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
#### What does this implement/fix? ####

The goal of this PR is to add a [vtki](https://github.com/vtkiorg/vtki)-based 3d backend to the visualization capabilitites of MNE.

#### Additional information ####

It is a work in progress, some features are still missing or incomplete as of now.

### CODE REVIEW

**Comment by larsoner** [comment]
This pull request **introduces 2 alerts** when merging 253a070f21fb2bf3836eee56049a1dcdd7287c95 into ece75432a4d51692dfc946fc7e6615be00b6a0dd - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-d48ea47ab395685a0decabdafbb8427c2c308f78)

**new alerts:**

* 1 for Unused import
* 1 for \_\_init\_\_ method returns a value

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging e26750c517305a26b8aa8f50dce33ac53e3d3e39 into ece75432a4d51692dfc946fc7e6615be00b6a0dd - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-9cab5bf6fcb4e647a08b9b649556f167245293d8)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by agramfort** [comment]
this works for me

```
import os.path as op

import mne
from mne.datasets import sample
from mne.viz.backends.renderer import set_3d_backend

# set_3d_backend("vispy")
# set_3d_backend("mayavi")
set_3d_backend("vtki")

print(__doc__)

data_path = sample.data_path()
subjects_dir = op.join(data_path, 'subjects')
raw_fname = op.join(data_path, 'MEG', 'sample', 'sample_audvis_raw.fif')
trans_fname = op.join(data_path, 'MEG', 'sample',
                      'sample_audvis_raw-trans.fif')
raw = mne.io.read_raw_fif(raw_fname)
trans = mne.read_trans(trans_fname)
src = mne.read_source_spaces(op.join(subjects_dir, 'sample', 'bem',
                                     'sample-oct-6-src.fif'))

mne.viz.plot_alignment(raw.info, trans=trans, subject='sample',
                       subjects_dir=subjects_dir, surfaces='head',
                       show_axes=True, dig=True, eeg=[], meg='sensors',
                       coord_frame='meg')
```

**Comment by GuillaumeFavelier** [comment]
Great news! Thanks for the feedback. 
You can also import the context manager `from mne.viz import use_3d_backend` and then do `with use_3d_backend('vtki'):`

I still need to check exactly why azure is not happy and add projection for montage and colormap support.

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 42e7d545930c8917ff0ffc4593e11c1d8ba95bfc into ece75432a4d51692dfc946fc7e6615be00b6a0dd - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-9cbb9011d8b7e6a2e8b8933d89aede38bd99f375)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging fc33e5b9c9d25d3bcc561a84c516d35826cf9efc into ece75432a4d51692dfc946fc7e6615be00b6a0dd - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-a2beec4fd5fc731dd646bcdce8767a94844f03a5)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by banesullivan** [comment]
Awesome work!!! This is really exciting to see you all incorporating `vtki` as a supporting module for 3D viz!

Feel free to ping me if you run into any issues getting this working. I see there's a timeout on one of your Travis builds... I poked around the build log along with your `vtki` code and all looks good, but let me know if you think that has to do with `vtki`.

**Comment by larsoner** [comment]
@banesullivan the Travis timeout might be spurious, I restarted it. The bigger problem is probably the Azure segfault (let me know your Azure / MS email if you can't access):

https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=560

> mne\viz\misc.py .                                                        [ 93%]
> ##[error]Cmd.exe exited with code '-1073741819'.
> ##[section]Finishing: Run tests

(The next test after `test_misc.py` is probably the `test_3d.py`)

We see something similar from Mayavi sometimes if we tweak tests, so it's probably really a VTK cleanup issue of some sort. Unfortunately `pytest-faulthandler` does not give us a traceback.

I'll try pushing a commit to remove `mayavi` tests to rule out that as the culprit. It could be that something about switching backends breaks things. (In my viz experience, garbage collection / GL contexts can be a problem, so our backend-switching mechanics, even if seemingly implemented properly, could have unpredictable negative consequences / segfaults.)

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 2056194835cc98a2c713dc06e8c0aff9fa817f94 into 32e1452b324f3542e7efcef9023b2222774f0785 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-1a07af3e66e50e91e1bcc9e895eebf07db1a14c3)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
Same error even with `mayavi` tests disabled:

https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=574

**Comment by banesullivan** [comment]
This is strang, can you use the verbose flag on that call to pytest that kicks off all the testing?

```
pytest -m "not ultraslowtest" mne --cov=mne -v
```

**Comment by banesullivan** [comment]
I've seen similar behavior before in our testing of `vtki` with not closing/cleaning up the `vtki.Plotter` object.

Also, in the `pytest` initialization output, it says:

```
pytest -m "not ultraslowtest" mne --cov=mne
========================== Starting Command Output ===========================
##[command]"C:\windows\system32\cmd.exe" /D /E:ON /V:OFF /S /C "CALL "D:\a\_temp\34be03c2-6dd7-4b57-8673-f74297ca8c0f.cmd""
============================= test session starts =============================
platform win32 -- Python 3.6.8, pytest-4.4.0, py-1.8.0, pluggy-0.9.0
rootdir: D:\a\1\s, inifile: setup.cfg
plugins: timeout-1.3.3, sugar-0.9.2, mock-1.10.1, faulthandler-1.5.0, cov-2.6.1
timeout: 30.0s
timeout method: thread
timeout func_only: False
collected 1113 items / 2 deselected / 1111 selected
```

Does that "`timeout: 30.0s`" phrase mean any test that takes longer than 30 seconds will fail? I don't think your test code could possibly take that long but this might also be worth looking into

I'll dig around in this PR's `vtki` code again and see if there's anything obvious

**Comment by banesullivan** [comment]
Do this doc-examples get tested and properly cleaned up?

https://github.com/GuillaumeFavelier/mne-python/blob/2056194835cc98a2c713dc06e8c0aff9fa817f94/mne/viz/misc.py#L836-L844

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 5897d349fa7f53f8a6cb7a6396e0bad1d01a2c32 into 2a23b0357adb7dfc837c592c03be597239bb196a - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-ec114a0393702e32f639c5be7fc37cbc27629600)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by GuillaumeFavelier** [comment]
I added the `-v` verbose flag to the `pytest` command so now I know that `test_backend_environment_setup()` passes but this one is just to test the configuration of the `MNE_3D_BACKEND` variable.

The real test is just after: `mne/viz/backends/tests/test_renderer.py::test_3d_backend` which creates a basic scene using the current 3d backend. Unfortunately, we don't have more infos as it seems to close unexpectingly.

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging dee33f9c17b2ac9bc002557de45d4e3ecc5b005b into 2a23b0357adb7dfc837c592c03be597239bb196a - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-d39bcffab9e95cb48befa231020708e2e37605f0)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging b4f4277024e6e8c18ec4e176bd2037a8e11758ac into 2c215d8f61e79f86a908da7f9e995819c5d62ddf - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-f392552b1b84b4c47cc906e7d6764231f348e68d)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 9fe42739fa7d946f046aa432956aabe48145e8c7 into 2c215d8f61e79f86a908da7f9e995819c5d62ddf - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-600efc40e3128d0f46393225c2911d5093fec3dd)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by GuillaumeFavelier** [comment]
I reproduce the same behaviour with pure `vtk` code. I will revert the code used for testing.

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging ab39fa12112dad1453c486144c7380114074c5b2 into 2c215d8f61e79f86a908da7f9e995819c5d62ddf - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-5b3d11f805c64bed18f852b4c35ac8ea61a454c6)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 659f900fe180763de27a7fe238647434af5c6047 into b3dfd7a1dd4cfc2fb501f9db9c4bcb3eb158cc2c - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-fc9687b8c462c543114515b82dd2650f622f636e)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: mne/viz/backends/_vtki.py [hunk 1]
```diff
@@ -191,12 +206,85 @@ def show(self):
     def close(self):
         self.plotter.close()
 
-    def set_camera(self, azimuth=None, elevation=None, distance=None,
-                   focalpoint=None):
-        return 0
+    def set_camera(self, azimuth=0.0, elevation=0.0, distance=1.0,
+                   focalpoint=(0, 0, 0)):
+        from numpy import cos, sin
```
Comment (agramfort) [inline mne/viz/backends/_vtki.py]: Don’t use nested import for numpy and use np prefix

**Comment by banesullivan** [comment]
I'm becoming more and more puzzled on how to get the Windows tests passing... Setting up a headless display isn't as straightforward as I'd have hoped.

See: https://help.appveyor.com/discussions/questions/36784-how-to-use-a-headless-display-windows

In the current tests with the Mayavi backend, do you all render and create 3D figures? Perhaps someone familiar with Mayavi could help figure out how to set up the headless display for VTK on Windows? Is there code within Mayavi that just handles this scenario?

**Comment by larsoner** [comment]
It should be doable. I have done it twice (at least), one for VisPy, which uses image-accuracy tests:

https://github.com/vispy/vispy/blob/master/appveyor.yml#L28

This was easy because you could tell VisPy which GL lib to use.

I then did it for Pyglet, which has no way to tell it which GL lib to use, since it just does `windll.opengl32`. This means you have to replace `C:\Windows\System32\opengl32.dll`. This was a nightmare to figure it out, but I eventually did, see:

https://github.com/LABSN/expyfun/blob/master/appveyor.yml#L19
https://github.com/LABSN/expyfun/blob/master/make/install_opengl.ps1#L14

**Comment by larsoner** [comment]
... and eventually you can set it up to see the proof that it uses the correct renderer, with whatever the VTK equivalent is of this GL renderer print check:
```
$ python -c "import pyglet; print(pyglet.gl.gl_info.get_renderer())"
Gallium 0.4 on llvmpipe (LLVM 3.4, 256 bits)
```
https://ci.appveyor.com/project/LABSN/expyfun/branch/master#L64

If you get a sad message about some Microsoft Software renderer 1.0 or so, it's not working yet...

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 69e14782165e145d2e74fec0b3792ecb52df5470 into b3dfd7a1dd4cfc2fb501f9db9c4bcb3eb158cc2c - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-6334d4c85de98fcbdf49217efac14d688c03978d)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by banesullivan** [comment]
> I then did it for Pyglet, which has no way to tell it which GL lib to use, since it just does windll.opengl32. This means you have to replace C:\Windows\System32\opengl32.dll. This was a nightmare to figure it out, but I eventually did, see:

@larsoner, you are the best! Got it working for `vtki` on AppVeyor!!! Now I suppose we just need to implement that here on Azure.

I simply copied/pasted that PowerShell script to replace the OpenGL library. Maybe we should add this to a CI-helper scripts repo somewhere? Is there one with more of a viz focus?

**Comment by larsoner** [comment]
I don't know of any with a viz focus, no. It seems a bit out of scope for `ci-helpers` since we're really talking about something specific to 3D libs that require OpenGL functionality on Windows CIs that don't have it (I have no similar problems on macOS or Linux, the latter thanks to `xvfb` but we could still automate this I guess).

I wouldn't want to foist maintenance responsibility for this sort of thing on the `astropy/ci-helpers` folks. Want to start a `vtkiorg` repo for it?

**Comment by banesullivan** [comment]
> I wouldn't want to foist maintenance responsibility for this sort of thing on the astropy/ci-helpers folks. Want to start a vtkiorg repo for it?

We can do it in `vtkiorg` - seems like an appropriate place. I'll get to adding it there

**Comment by banesullivan** [comment]
I piggy-backed off of `astropy/ci-helpers` and created [`vtkiorg/ci-helpers`](https://github.com/vtkiorg/ci-helpers/tree/master). Everything is live and working for `vtki`'s tests on AppVeyor now. @larsoner, I sent you an invite to join `vtkiorg` - could you help me make sure the script's licensing and all are correct since I simply copy/pasted it

**Comment by agramfort** [comment]
To avoid confusion I would call this repo gl-ci-helpers

**Comment by larsoner** [comment]
MIT is fine by me, agreed a non-overlapping name would help, especially since every place I use the GL helpers I will probably use `astropy/ci-helpers`, too, and it would be nice not to have to tell it to clone into a different dir

**Comment by banesullivan** [comment]
Fair point! Renamed to `gl-ci-helpers`

**Comment by banesullivan** [comment]
Will that script work for the Azure windows builds in this PR?

**Comment by larsoner** [comment]
I'll push a commit to try

**Comment by banesullivan** [comment]
Also, I'm thinking about future development from `vtki` and the implications that might have on dependent projects like `mne-python` after this PR. 

Does anyone here have experience setting up downstream testing pipelines? Ideally, I'd like to run our internal tests for `vtki` then trigger a series of tests to make sure we don't break features for external projects like `mne-python` - is there a CI service that does this?

**Comment by larsoner** [comment]
Not that I know of. We're happy to open issues upstream if things break :)

**Comment by larsoner** [comment]
It did not work, in retrospect it makes sense:
```
SUCCESS: The file (or folder): "C:\Windows\system32\opengl32.dll" now owned by the administrators group.
APPVYR-WIN\appveyor: No mapping between account names and security IDs was done.
Successfully processed 0 files; Failed processing 1 files
Remove-item : Cannot remove item C:\Windows\system32\opengl32.dll: Access to the path 
'C:\Windows\system32\opengl32.dll' is denied.
At D:\a\1\s\gl-ci-helpers\appveyor\install_opengl.ps1:26 char:5
+     Remove-item -LiteralPath $filepath
+     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (C:\Windows\system32\opengl32.dll:FileInfo) [Remove-Item], Unauthorize 
   dAccessException
    + FullyQualifiedErrorId : RemoveFileSystemItemUnAuthorizedAccess,Microsoft.PowerShell.Commands.RemoveItemCommand
Downloading https://github.com/vispy/demo-data/raw/master/mesa/opengl32_mingw_64.dll
```
For Azure we'll need to change the user that gets permissions. Maybe there is some powershell command to automagically get it so the AppVeyor/Azure can stay unified.

**Comment by larsoner** [comment]
Probably something like this

https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.localaccounts/get-localgroupmember?view=powershell-5.1

**Comment by banesullivan** [comment]
I'm clueless when it comes to PowerShell commands and Windows stuff... not sure how to fix this...

**Comment by larsoner** [comment]
I can. Can you enable AppVeyor and Azure for `gl-ci-helpers`? I'll test there via PR

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 7c32201aef613ba9ddfc91a4df4b5cd85561053f into b3dfd7a1dd4cfc2fb501f9db9c4bcb3eb158cc2c - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-baa7982435c034b8ffd3b7e4f8ca23d8092b49f8)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
[Windows CIs fixed](https://dev.azure.com/mne-tools/mne-python/_build/results?buildId=621), @GuillaumeFavelier feel free to complete the backend now :)

**Comment by GuillaumeFavelier** [comment]
Amazing what happened in here. thank you @larsoner, @banesullivan and @agramfort

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 0c134cf20902181715e9c08cc964d63b82924e7a into b3dfd7a1dd4cfc2fb501f9db9c4bcb3eb158cc2c - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-f68b40cc7147d9e820ab3710451a2db210a217d1)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by banesullivan** [comment]
@GuillaumeFavelier, Do you all need a tagged release from `vtki` for this? Asking because we made a few big fixes from your work here

**Comment by GuillaumeFavelier** [comment]
I wouldn't say we need it because we can fixe the commit hash in the `requirements.txt` and `environment.yml` but that would be appreciated of course as it is easier to manage.

**Comment by banesullivan** [comment]
Okay, I'm going to hold off on tagging a release to mitigate a few issues and make sure our all is good on some recent changes that might affect deployment

**Comment by larsoner** [comment]
@GuillaumeFavelier if you can rebase and add a RuntimeError + test skip for the snapshot montage for vtki for now then we can merge

**Comment by GuillaumeFavelier** [comment]
okay, working on it.

**Comment by banesullivan** [comment]
FYI: as soon as https://github.com/vtkiorg/vtki/pull/190 is tested on a Windows machine, I'll merge and tag `v0.18.4` for `vtki` if you all need to set a dependency for `vtki`

**Comment by larsoner** [comment]
This pull request **introduces 1 alert** when merging 80aa30ca5d1d5dad0afc1779ba9b8827d4da74ce into 9cc5d1457c29151eea73fa7f14619724182960c1 - [view on LGTM.com](https://lgtm.com/projects/g/mne-tools/mne-python/rev/pr-0270ced7426515eb1814a4e1366fe425e8795e56)

**new alerts:**

* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by larsoner** [comment]
Thanks @GuillaumeFavelier

**Pull request merged**